libs/error-stack/src/sink.rs

```diff
@@ -0,0 +1,645 @@
+                eprintln!("ReportSink was dropped without being consumed");
+            }
+            BombState::Defused => {}
+        }
+    }
+}
+/// A sink for collecting multiple [`Report`]s into a single [`Result`].
+///
+/// [`ReportSink`] allows you to accumulate multiple errors or reports and then
+/// finalize them into a single `Result`. This is particularly useful when you
+/// need to collect errors from multiple operations before deciding whether to
+/// proceed or fail.
+///
+/// The sink is equipped with a "bomb" mechanism to ensure proper usage,
+/// if the sink hasn't been finished when dropped, it will emit a warning or panic,
+/// depending on the constructor used.
+///
+/// # Examples
+///
+/// ```
+/// use error_stack::{ReportSink, Result};
+///
+/// #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+/// struct InternalError;
+///
+/// impl core::fmt::Display for InternalError {
+///     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+///         f.write_str("Internal error")
+///     }
+/// }
+///
+/// impl core::error::Error for InternalError {}
+///
+/// fn operation1() -> Result<u32, InternalError> {
+///     // ...
+///     # Ok(42)
+/// }
+///
+/// fn operation2() -> Result<(), InternalError> {
+///     // ...
+///     # Ok(())
+/// }
+///
+/// fn process_data() -> Result<(), [InternalError]> {
+///     let mut sink = ReportSink::new();
+///
+///     if let Some(value) = sink.attempt(operation1()) {
+///         // process value
+///         # let _value = value;
+///     }
+///
+///     if let Err(e) = operation2() {
+///         sink.add(e);
+///     }
+///
+///     sink.finish()
+/// }
+/// # let _result = process_data();
+/// ```
+#[must_use]
+pub struct ReportSink<C> {
+    report: Option<Report<[C]>>,
+    bomb: Bomb,
+}
+
+impl<C> ReportSink<C> {
+    /// Creates a new [`ReportSink`].
+    ///
+    /// If the sink hasn't been finished when dropped, it will emit a warning.
+    pub const fn new() -> Self {
+        Self {
+            report: None,
+            bomb: Bomb::warn(),
+        }
+    }
+
+    /// Creates a new [`ReportSink`].
+    ///
+    /// If the sink hasn't been finished when dropped, it will panic.
+    pub const fn new_armed() -> Self {
+        Self {
+            report: None,
+            bomb: Bomb::panic(),
+        }
+    }
+
+    /// Adds a [`Report`] to the sink.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use error_stack::{ReportSink, Report};
+    /// # use std::io;
+    /// let mut sink = ReportSink::new();
+    /// sink.add(Report::new(io::Error::new(
+    ///     io::ErrorKind::Other,
+    ///     "I/O error",
+    /// )));
+    /// ```
+    pub fn add(&mut self, report: impl Into<Report<[C]>>) {
+        let report = report.into();
+
+        match self.report.as_mut() {
+            Some(existing) => existing.append(report),
+            None => self.report = Some(report),
+        }
+    }
```

Comment:
`add` is a term which is rarely used in Rust. Maybe we want to use `append` or the `Extend` trait? I don't have a strong opinion on that, I just want to mention it 🙂

Reply:
Thought about that as well but didn't rename it because of the follow up PR that applies the changes to the worktree. I was thinking of renaming it to either "push" or "append" (more likely append) to be in line with the existing methods in Report<[C]>.

What di you think about append?

Reply:
As I proposed `append` myself I'm fine with that 🚀 

We can always change it later, no rush 👍🏻